subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultTaskOutputFilesRepository.java

```diff
@@ -0,0 +1,113 @@
+package org.gradle.api.internal.changedetection.state;
+
+import org.gradle.api.invocation.Gradle;
+import org.gradle.cache.CacheBuilder;
+import org.gradle.cache.CacheRepository;
+import org.gradle.cache.PersistentCache;
+import org.gradle.cache.PersistentIndexedCache;
+import org.gradle.cache.PersistentIndexedCacheParameters;
+import org.gradle.cache.internal.FileLockManager;
+import org.gradle.internal.nativeintegration.filesystem.FileType;
+import org.gradle.util.GradleVersion;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.util.Collections;
+
+import static org.gradle.cache.internal.filelock.LockOptionsBuilder.mode;
+
+public class DefaultTaskOutputFilesRepository implements TaskOutputFilesRepository, Closeable {
+
+    private final static String CACHE_DISPLAY_NAME = "Build Output Cleanup Cache";
+
+    private final CacheRepository cacheRepository;
+    private final PersistentCache cacheAccess;
+    private final FileSystemMirror fileSystemMirror;
+    private final PersistentIndexedCache<String, Boolean> outputFiles; // The value is true if it is an output file, false if it is a parent of an output file
+
+    public DefaultTaskOutputFilesRepository(CacheRepository cacheRepository, Gradle gradle, FileSystemMirror fileSystemMirror, InMemoryCacheDecoratorFactory inMemoryCacheDecoratorFactory) {
+        this.cacheRepository = cacheRepository;
+        this.cacheAccess = createCache(gradle);
+        this.fileSystemMirror = fileSystemMirror;
+        this.outputFiles = cacheAccess.createCache(cacheParameters(inMemoryCacheDecoratorFactory));
+    }
+
+    private static PersistentIndexedCacheParameters<String, Boolean> cacheParameters(InMemoryCacheDecoratorFactory inMemoryCacheDecoratorFactory) {
+        return new PersistentIndexedCacheParameters<String, Boolean>("outputFiles", String.class, Boolean.class)
+            .cacheDecorator(inMemoryCacheDecoratorFactory.decorator(100000, true));
+    }
+
+    @Override
+    public boolean isGeneratedByGradle(File file) {
+        File absoluteFile = file.getAbsoluteFile();
+        return containsFilesGeneratedByGradle(absoluteFile) || isContainedInAnOutput(absoluteFile);
+    }
+
+    private Boolean isContainedInAnOutput(File file) {
+        File currentFile = file;
+        do {
+            if (outputFiles.get(currentFile.getPath()) == Boolean.TRUE) {
+                return true;
+            }
+            currentFile = currentFile.getParentFile();
+        } while (currentFile != null);
+        return false;
+    }
+
+    private boolean containsFilesGeneratedByGradle(File file) {
+        return outputFiles.get(file.getPath()) != null;
+    }
+
+    @Override
+    public void recordOutputs(TaskExecution taskExecution) {
+        for (String outputFilePath : taskExecution.getDeclaredOutputFilePaths()) {
+            FileSnapshot fileSnapshot = fileSystemMirror.getFile(outputFilePath);
+            File outputFile = new File(outputFilePath);
```

Comment:
We could split the path at the file separator and then re-join it again. I doubt that this would be much faster and I didn't see any Hotspots for creating these File objects in the profiler.